modules/invenio-oaiharvester/invenio_oaiharvester/harvester.py

```diff
@@ -1874,9 +1246,251 @@ def map(self):
+                    if isinstance(current_temp, list):
+                        current_temp = current_temp[0]
+                        target = target[0]
+                elif i_r not in current_temp:
+                    current_temp[i_r] = [] if '[]' in i else {}
+                    if isinstance(current_temp[i_r], list):
+                        current_temp[i_r] = target[i_r]
+                    else:
+                        current_temp[i_r].update(target[i_r])
+                    break
+            return result
+
+        def merge_data_by_mapping_keys(parent_key, dic):
+            """Merge all sub data by parent key(prefix key)."""
+            current_key = parent_key
+            if isinstance(dic, dict):
+                for key, val in dic.items():
+                    full_key = current_key + '.' + key
+                    if full_key in lst_keys_unique:
+                        if isinstance(val, str):
+                            val = {'#text': val}
+                        if dict_data.get(full_key, None):
+                            dict_data[full_key].append(val)
+                        else:
+                            dict_data[full_key] = [val]
+                    else:
+                        merge_data_by_mapping_keys(full_key, val)
+            elif isinstance(dic, list):
+                for i in dic:
+                    for key, val, in i.items():
+                        full_key = current_key + '.' + key
+                        if full_key in lst_keys_unique:
+                            if isinstance(val, str):
+                                val = {'#text': val}
+                            if dict_data.get(full_key, None):
+                                dict_data[full_key].append(val)
+                            else:
+                                dict_data[full_key] = [val]
+                        else:
+                            merge_data_by_mapping_keys(full_key, val)
+            elif isinstance(dic, str) and '@' not in current_key:
+                dic = {'#text': dic}
+                if dict_data.get(current_key, None):
+                    dict_data[current_key].append(dic)
+                else:
+                    dict_data[current_key] = [dic]
+
+        def parse_to_obj_data_by_mapping_keys(vals, keys):
+            """Parse all data to type of object by mapping key."""
+            try:
+                list_result = []
+                root_key = ''
+                for val_obj in vals:
+                    temp_lst = []
+                    for key_pair in keys:
+                        for mapping_key, item_sub_key in key_pair.items():
+                            for item_sub_key_form in item_sub_keys_form:
+                                if item_sub_key_form.replace("[]", "") \
+                                        == item_sub_key:
+                                    item_sub_key = item_sub_key_form
+                                    break
```

Comment:
There are four loops nested.
Please split this source code into the small methods.